File: frontend/src/api/schemas/collection.ts
import { z } from 'zod'
import { CategoryIdSchema, NoteAuthorSchema } from './note'

/** A note as it appears inside a collection (mirror of backend CollectionNoteOut). */
export const CollectionNoteSchema = z.object({
  id: z.string(),
  title: z.string(),
  summary: z.string(),
  category: CategoryIdSchema,
  createdAt: z.string(),
  readMinutes: z.number().int().positive(),
  author: NoteAuthorSchema,
})
export type CollectionNote = z.infer<typeof CollectionNoteSchema>

/** Mirror of backend CollectionOut. */
export const CollectionSchema = z.object({
  id: z.string(),
  title: z.string(),
  description: z.string().default(''),
  entryCount: z.number().int().nonnegative().default(0),
  createdAt: z.string(),
  updatedAt: z.string(),
})
export type Collection = z.infer<typeof CollectionSchema>

export const CollectionListSchema = z.array(CollectionSchema)

/** Mirror of backend CollectionDetailOut (collection + ordered entries). */
export const CollectionDetailSchema = CollectionSchema.extend({
  entries: z.array(CollectionNoteSchema).default([]),
})
export type CollectionDetail = z.infer<typeof CollectionDetailSchema>

/** GET /notes/{id}/collection — sidebar context. */
export const NoteCollectionContextSchema = z.object({
  collection: CollectionSchema,
  entries: z.array(CollectionNoteSchema),
  currentIndex: z.number().int(),
})
export type NoteCollectionContext = z.infer<typeof NoteCollectionContextSchema>

export type CollectionCreateIn = { title: string; description?: string }
export type CollectionUpdateIn = { title?: string; description?: string }

File: backend/app/schemas/collection.py
"""Collection / sidebar schemas."""
from __future__ import annotations

from pydantic import Field

from app.schemas._base import CamelModel, UtcDateTime
from app.schemas.note import CategoryId, NoteAuthorOut


class CollectionCreateIn(CamelModel):
    title: str = Field(min_length=1, max_length=255)
    description: str = Field(default="", max_length=4000)


class CollectionUpdateIn(CamelModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=4000)


class CollectionEntryAddIn(CamelModel):
    note_id: str
    sort_order: int | None = Field(default=None, ge=0)


class CollectionReorderIn(CamelModel):
    """Full ordered list of note ids for a collection (drag-sort)."""

    note_ids: list[str]


class CollectionNoteOut(CamelModel):
    id: str
    title: str
    summary: str
    category: CategoryId
    created_at: UtcDateTime
    read_minutes: int = Field(ge=1)
    author: NoteAuthorOut


class CollectionOut(CamelModel):
    id: str
    title: str
    description: str = ""
    entry_count: int = Field(default=0, ge=0)
    created_at: UtcDateTime
    updated_at: UtcDateTime


class CollectionDetailOut(CollectionOut):
    entries: list[CollectionNoteOut] = Field(default_factory=list)


class NoteCollectionContextOut(CamelModel):
    collection: CollectionOut
    entries: list[CollectionNoteOut] = Field(default_factory=list)
    current_index: int = Field(ge=0)